Write about best-practice high availability and scaling of cert-manager components (#1330)
cert-manager/website: content/docs/installation/best-practice.md

You may find that the default resources do not comply with the security policy on your Kubernetes cluster
and in that case you can modify the installation configuration using Helm chart values to override the defaults.

## High Availability

cert-manager has three long-running components: controller, cainjector, and webhook.
Each of these components has a Deployment and by default each Deployment has 1 replica
but this does not provide high availability.
The Helm chart for cert-manager has parameters to configure the `replicaCount` for each Deployment,
and in production you should use 2 or more replicas to achieve high availability.

> 

Reviewer: ```yaml
replicaCount: 2
webhook:
  replicaCount: 3
cainjector:
  replicaCount: 2
```

Author: Done.

### controller and cainjector

The controller and cainjector components use [leader election](https://pkg.go.dev/k8s.io/client-go/tools/leaderelection)
to ensure that only one replica is active.
This prevents conflicts which would arise if multiple replicas were reconciling the same API resources.
So in these components you can use multiple replicas to achieve high availability but not for horizontal scaling.

Use two replicas to ensures that there is a standby Pod scheduled to a Node which is ready to take leadership,
should the current leader encounter a disruption.

> For example, if the Node on which the leader Pod is running is drained.

Reviewer 2: ```suggestion
For example, when the leader Pod is drained from its node.
```

Or, if the leader Pod encounters an unexpected deadlock.

There is little justification for using more than 2 replicas of these components.
Further replicas *may* add a degree of resilience
if you have the luxury of sufficient Nodes

> with sufficient CPU and memory to accommodate the standby replicas.

Reviewer 2: ```suggestion
with sufficient CPU and memory to accommodate additional standby replicas.
```

### webhook

By default the cert-manager webhook Deployment has 1 replica, but in production you should use 3 or more.
If the cert-manager webhook is unavailable, all API operations on cert-manager custom resources will fail,

> and this will disrupt any software that creates, updates or deletes cert-manager custom resources.

Reviewer 2: It might also generally disrupt the cluster -- certainly GKE yelled at me about this, I think.

Author: Done.

for each of the components.
For example, the following Helm chart values add topology spread constraints for all three long-running components,
to request (but not require) Kubernetes to avoid scheduling Pods of the same Deployment to the same zone or node.

```yaml
topologySpreadConstraints:
- maxSkew: 1
  topologyKey: topology.kubernetes.io/zone
  whenUnsatisfiable: ScheduleAnyway
  labelSelector:
    matchLabels:
      app.kubernetes.io/instance: cert-manager
      app.kubernetes.io/component: controller
- maxSkew: 1
  topologyKey: kubernetes.io/hostname
  whenUnsatisfiable: ScheduleAnyway
  labelSelector:
    matchLabels:
      app.kubernetes.io/instance: cert-manager
      app.kubernetes.io/component: controller

webhook:
  replicaCount: 3
  topologySpreadConstraints:
  - maxSkew: 1
    topologyKey: topology.kubernetes.io/zone
    whenUnsatisfiable: ScheduleAnyway
    labelSelector:
      matchLabels:
        app.kubernetes.io/instance: cert-manager
        app.kubernetes.io/component: webhook
  - maxSkew: 1
    topologyKey: kubernetes.io/hostname
    whenUnsatisfiable: ScheduleAnyway
    labelSelector:
      matchLabels:
        app.kubernetes.io/instance: cert-manager

>         app.kubernetes.io/component: webhook

Reviewer: I'm easy here, not having it is less YAML.
Having it means it's explicit.

>         app.kubernetes.io/component: webhook

Reviewer 3: I think if we can verify that this works by default, we should not tell people to configure this (because we want to keep our docs as simple as possible). Instead, we can just link to the documentation you shared and say that this works correctly by default.

>         app.kubernetes.io/component: webhook

Author: I agree that according to the docs the desired anti-affinity scheduling should happen by default.
I've updated this paragraph.

cainjector:
  replicaCount: 2
  topologySpreadConstraints:
  - maxSkew: 1
    topologyKey: topology.kubernetes.io/zone
    whenUnsatisfiable: ScheduleAnyway
    labelSelector:
      matchLabels:
        app.kubernetes.io/instance: cert-manager
        app.kubernetes.io/component: cainjector
  - maxSkew: 1
    topologyKey: kubernetes.io/hostname
    whenUnsatisfiable: ScheduleAnyway
    labelSelector:
      matchLabels:
        app.kubernetes.io/instance: cert-manager
        app.kubernetes.io/component: cainjector
```

### PodDisruptionBudget

For high availability you should also deploy a `PodDisruptionBudget` resource,
with `minAvailable=1` *or* with `maxUnavailable=1`.

> This ensures that a *voluntary* disruption, such as the draining of a Node, can not proceed

Reviewer 2: ```suggestion
This ensures that a *voluntary* disruption, such as the draining of a Node, cannot proceed
```

until at least one other replica has been successfully scheduled and started on another Node.
The Helm chart has parameters to enable and configure a PodDisruptionBudget
for each of the long-running cert-manager components.

> 

Reviewer: Perhaps include the relevant subset of YAML to configure PDBs? As you have done with the topology settings?

```yaml
podDisruptionBudget:
  enabled: true
  minAvailable: 1
webhook:
  podDisruptionBudget:
    enabled: true
    minAvailable: 1
cainjector:
  podDisruptionBudget:
    enabled: true
    minAvailable: 1
```

Author: Done.

## Scalability

cert-manager has three long-running components: controller, cainjector, and webhook.
The Helm chart does not include resource requests and limits for any of these,
so you should supply resource requests and limits which are appropriate for your cluster.

### controller and cainjector

The controller and cainjector components use leader election to ensure that only one replica is active.
This prevents conflicts which would arise if multiple replicas were reconciling the same API resources.

> You can not use horizontal scaling for these components.

Reviewer 2: ```suggestion
You cannot use horizontal scaling for these components.
```

Use vertical scaling instead.

#### Memory

Use vertical scaling to assign sufficient memory resources to these components.
The memory requirements will be higher on clusters with very many API resources or with large API resources.
This is because each of the components reconciles one or more Kubernetes API resources,
and each component will cache the metadata and sometimes the entire resource in memory,
so as to reduce the load on the Kubernetes API server.

> For example, if the cluster contains very many CertificateRequest resources,

Reviewer: Perhaps a use case is more illustrative. Eg,

```suggestion
If your cluster contains a high volume of `CertificateRequest` resources such as when using many ephemeral or short lived certificate rotates frequently,
```

Unless I have missed the point here.

> you will need to increase the memory limit of the controller Pod.

Reviewer: Is it best practice to limit this to only the installation namespace?

> you will need to increase the memory limit of the controller Pod.

Author: I suppose it's not technically "best practice" but I've added it anyway. 
See what you think.

#### CPU

Use vertical scaling to assign sufficient CPU resources to the these components.
The CPU requirements will be higher on clusters where there are very frequent updates to the resources which are reconciled by these components.
Whenever a resource changes, it will be queued to be re-reconciled by the component.
Higher CPU resources allow the component to process the queue faster.

#### Workers

Each of these component spawns a pool of worker threads to reconcile the queue of API resources.
You may need to increase this if you increase the CPU resources,
so that each component can make efficient use of the CPU cores assigned to it.

By default the controller [uses 5 workers per controller](https://github.com/cert-manager/cert-manager/blob/3b0a5cec4140e92ba12f3eace362a4bd65fbb30e/cmd/controller/app/options/options.go#L180-L181)
which should be adequate for most clusters.
You can change the number of workers using the `--concurrent-workers` [flag of the controller manager](../cli/controller.md).

The cainjector uses a single worker for each of the resource types that it reconciles,

> ⚠️ This this is not yet configurable.

> 

Reviewer: I think you are right in that it's not "best practice" to change this setting. It's more of an operational concern at a considerable scale. So I'd :+1: removing it from this document.

You may find that the default resources do not comply with the security policy on your Kubernetes cluster
and in that case you can modify the installation configuration using Helm chart values to override the defaults.

## High Availability

cert-manager has three long-running components: controller, cainjector, and webhook.
Each of these components has a Deployment and by default each Deployment has 1 replica
but this does not provide high availability.
The Helm chart for cert-manager has parameters to configure the `replicaCount` for each Deployment.
In production we recommend the following `replicaCount` parameters:

```yaml
replicaCount: 2
webhook:

>   replicaCount: 3

Reviewer 3: I think a user might want even more webhook replicas, maybe we should explain why we set the replicas for leader-elected controllers to 2 and for non-leader elected controllers to >= 3?

cainjector:
  replicaCount: 2
```

### controller and cainjector

The controller and cainjector components use [leader election](https://pkg.go.dev/k8s.io/client-go/tools/leaderelection)
to ensure that only one replica is active.
This prevents conflicts which would arise if multiple replicas were reconciling the same API resources.
So in these components you can use multiple replicas to achieve high availability but not for horizontal scaling.

Use two replicas to ensures that there is a standby Pod scheduled to a Node which is ready to take leadership,
should the current leader encounter a disruption.
For example, when the leader Pod is drained from its node.
Or, if the leader Pod encounters an unexpected deadlock.

There is little justification for using more than 2 replicas of these components.

> Further replicas *may* add a degree of resilience
> if you have the luxury of sufficient Nodes
> with sufficient CPU and memory to accommodate additional standby replicas.

Reviewer 3: ```suggestion
```

I don't think there is a real reason here, we are basically saying that you can run more than 2 but only if you like to waste compute resources.

Author: It would help if you were unlucky enough to simultaneously lose two data centers or if you lose  two nodes of a single zone cluster...but I think might remove it for the sake of making the document shorter and easy to digest.

### webhook

By default the cert-manager webhook Deployment has 1 replica, but in production you should use 3 or more.
If the cert-manager webhook is unavailable, all API operations on cert-manager custom resources will fail,

> and this will disrupt any software that creates, updates or deletes cert-manager custom resources,

Reviewer 3: ```suggestion
and this will disrupt any software that creates, updates or deletes cert-manager custom resources (including cert-manager itself),
```

      app.kubernetes.io/instance: cert-manager
      app.kubernetes.io/component: controller

webhook:
  topologySpreadConstraints:
  - maxSkew: 1
    topologyKey: topology.kubernetes.io/zone
    whenUnsatisfiable: ScheduleAnyway
    labelSelector:
      matchLabels:
        app.kubernetes.io/instance: cert-manager
        app.kubernetes.io/component: webhook
  - maxSkew: 1
    topologyKey: kubernetes.io/hostname
    whenUnsatisfiable: ScheduleAnyway
    labelSelector:
      matchLabels:
        app.kubernetes.io/instance: cert-manager
        app.kubernetes.io/component: webhook

cainjector:
  topologySpreadConstraints:
  - maxSkew: 1
    topologyKey: topology.kubernetes.io/zone
    whenUnsatisfiable: ScheduleAnyway
    labelSelector:
      matchLabels:
        app.kubernetes.io/instance: cert-manager
        app.kubernetes.io/component: cainjector
  - maxSkew: 1
    topologyKey: kubernetes.io/hostname
    whenUnsatisfiable: ScheduleAnyway
    labelSelector:
      matchLabels:
        app.kubernetes.io/instance: cert-manager
        app.kubernetes.io/component: cainjector
```

### PodDisruptionBudget

For high availability you should also deploy a `PodDisruptionBudget` resource,

> with `minAvailable=1` *or* with `maxUnavailable=1`.

Reviewer 3: I'm not sure about `maxUnavailable=1`, what happens if you only have 1 replica running? Is it OK to remove that replica in that situation?
Also, if you have a lot of replicas, `maxUnavailable > 1` might be ok too.

Reviewer 2: `maxUnavailable: 1` requires you to have 2 running, otherwise you're stuck. Speaking from experience.

Author: Both settings require at least 2 replicas which is why Kyverno omit the PodDisruptionBudget if Replicas < 2

https://github.com/kyverno/kyverno/blob/53eab3d777d3dc33a1cfac73ea069c910e4b3f45/charts/kyverno/templates/background-controller/poddisruptionbudget.yaml#L1-L16

I've added a warning note about this.

I think `maxUnavailable > 1` might ok for the webhook if you have 3 replicas and you want to upgrade Nodes 2 at a time.
I'll just recommend `minAvailable=1` for the sake of simplicity.
